### ISSUE TITLE
magit-rebase: check for commutativity?

### ISSUE THREAD

**Issue by xmo-odoo** [magit/magit#4530]
`magit-rebase` is absolutely stellar, however one issue it shares with the original `rebase` is that it doesn't understand *commutativity*. As such it can be easy to miss dependencies between patches, swap patches which should not be and end up in conflict hell. AFAIK at that point it's out of magit's hands and the only option is to cancel and redo the entire rebase configuration.

Magit warning a user that their rebase sequence will conflict before actually running the rebase would be tremendously useful, though I don't know if it's possible as-is, at least without significant side-effects.

**Comment by tarsius** [comment]
> `magit-rebase` is absolutely stellar, however one issue it shares with the original `rebase` is that it doesn't understand _commutativity_. As such it can be easy to miss dependencies between patches, swap patches which should not be and end up in conflict hell. AFAIK at that point it's out of magit's hands and the only option is to cancel and redo the entire rebase configuration.

Yes, that's beyond the scope of Magit--I mean they write whole new vcs just to deal with that...

> Magit warning a user that their rebase sequence will conflict before actually running the rebase would be tremendously useful, though I don't know if it's possible as-is, at least without significant side-effects.

It would be a big undertaking and essentially only so that Magit can say "There's gonna be a conflict. Do you want to deal with it?" instead of Git potentially saying "There's a conflict. Deal with it." At that point the user can still abort and all is good again. So no, I don't think detecting a conflict beforehand is worth the effort and risk of false flags.

**Comment by xmo-odoo** [comment]
> > `magit-rebase` is absolutely stellar, however one issue it shares with the original `rebase` is that it doesn't understand _commutativity_. As such it can be easy to miss dependencies between patches, swap patches which should not be and end up in conflict hell. AFAIK at that point it's out of magit's hands and the only option is to cancel and redo the entire rebase configuration.
> 
> Yes, that's beyond the scope of Magit--I mean they write whole new vcs just to deal with that...

They wrote a whole new vcs based entirely on the concept, which is a rather different proposition than checking whether permutations are valid in a restricted context.

> > Magit warning a user that their rebase sequence will conflict before actually running the rebase would be tremendously useful, though I don't know if it's possible as-is, at least without significant side-effects.
> 
> It would be a big undertaking and essentially only so that Magit can say "There's gonna be a conflict. Do you want to deal with it?" instead of Git potentially saying "There's a conflict. Deal with it."

Rather that Magit would be able to say "There's gonna be a conflict, are you sure that is what you want?" because IME conflicts during rebases are almost always mistakes in the reordering and folding of commits.

> At that point the user can still abort and all is good again.

Except for having to re-do the entire rebase sequencing again which can be involved when cleaning up complicated branches, unless (ma)git saves the rebase-configuration buffer somewhere from which it can easily be restored on retry?

**Comment by tarsius** [comment]
> Except for having to re-do the entire rebase sequencing again which can be involved when cleaning up complicated branches, 

I can see how it would be nice to know about conflicts before hand in such cases. I usually perform multiple smaller rebases instead of one complex one and would recommend other do the same, but I can see that in some cases it might be feasible to do that.

There's an official (by @kyleam) Magit extension for `git imerge`: https://github.com/magit/magit-imerge. I don't actually think it will help here because it doesn't seem to provide a `rebase --interactive` variant.

Never-the-less it might be useful to watch the (short) video if only to get a visualization of what it means to determine if there is a conflict between two lines of development. (A reordered sequence of commits can be thought of as such a line of development, though you would have to inject additional revert commits to complete the picture.) It isn't enough to compare the tips of the two branches obviously, those might very well be conflict free, in which case a *merge* is trivial, but for a rebase many more comparisons would have to be made, and the `git-imerge` introductory video does a good job visualizing that.

If Magit were to attempt to determine before hand if there are any conflicts it would have to perform all step-by-step merges up front. This isn't trivial, `git-imerge` is several thousand lines of code.

I don't think `git-imerge` can actually be used for reordering commits as it expects two actual branches, while the (reordered) sequence of commits in a `git-rebase-todo` file/buffer is more of a wishlist. One could pressingly create an actual commit sequence from that, but that would involve injecting revert commits... and at that point we are implementing something that approaches the complexity of `git rebase` itself.

> unless (ma)git saves the rebase-configuration buffer somewhere from which it can easily be restored on retry?

That would be another, likely more feasible, approach, but it's also not trivial. I recommend you implement something like that for your own use and maybe publish it as a separate package for others to use.

This is how I imagine that could work: Whenever a rebase brings up a `git-rebase-todo` buffer, you somehow record the respective file but also other related information such as the arguments that rebase was called with. Then when `git-rebase-todo` pops up again, you compare the arguments that were used this time around, which the arguments of past invocations and if a match is found, then you give the user the option of replacing the pristine rebase sequence with the recorded sequence for further editing.

In conclusion, I think this is very complex (not just the underlying logic but also the ui) and I still do not want to implement it, but it is certainly possible to implement improvements and I encourage you to have a stab at it. While I do not rule it out, I do however suspect that I won't want to merge the result into Magit.
